[Spring 경로 조회 - 2단계] 정(정진우) 미션 제출합니다.

bugoverdose opened the pull request:
안녕하세요. 정입니다.

이전 PR의 리뷰 주신 분들 참고해서, 요구사항 구현 및 리팩토링했습니다.

제자리걸음을 하고 있는 느낌이 들어서 조금 일찍 리뷰 요청 보냅니다.

bugoverdose committed:
test(study): check jgrapht library

- learn how to use custom edge for finding shortestPath to get edge list

bugoverdose committed:
docs(README): add fair policy

bugoverdose committed:
feat(domain): implement Path2 with jgrapht library

- add tests

bugoverdose committed:
refactor(service): replace PathManager and Path with Path2 domain

bugoverdose committed:
refactor: replace old Path with Path2 and change domain name

- delete unused classes

bugoverdose committed:
refactor(dto): remove DtoAssembler and rollback

bugoverdose committed:
feat: add extraFare field at db and update CRUD for line

- change all dto, domain, entity regarding line
- fix all broken tests

bugoverdose committed:
test(dao): add tests on line update

bugoverdose committed:
feat(domain): add validation for line extra fare value

- throw IllegalArgumentException if extra fare is under 0

bugoverdose committed:
test(service): add test on extra fare validation at create and update line

- add missing tests on update line

bugoverdose committed:
test(ui): add missing tests on extra fare validation

- fix wrong tests

bugoverdose committed:
refactor: minor cleansing

bugoverdose committed:
docs(README): update feature list

bugoverdose committed:
feat(dao): implement findAllByIds at LineDao

- add test case

bugoverdose committed:
feat(repository): implement findAllLinesByIds at SubwayRepository

- add test case

bugoverdose committed:
feat(domain): implement BasicFare and DistanceOverFare

- implement decorator pattern
- add test cases

bugoverdose committed:
feat(domain): implement LineOverFare as decorator

- add test cases

bugoverdose committed:
feat(domain): implement AgeDiscountFare as decorator

- add test cases

bugoverdose committed:
refactor(domain): replace old fare package with new fare2

- fix and update use cases

bugoverdose committed:
docs(README): add features on elderly and baby discount

bugoverdose committed:
feat(service): implement age discount and line extra fare calculation

bugoverdose committed:
feat(domain): implement Navigator and NavigatorJgraphtAdapter domain

- implement adapter pattern to separate jgrapht library from Path domain
- add test cases

bugoverdose committed:
feat(domain): implement new Path domain with NavigatorJgraphtAdapter

- add test cases

bugoverdose committed:
refactor: replace old Path with new Path domain

bugoverdose committed:
refactor(domain): separate fare and path domains

- delete unused files

bugoverdose committed:
feat(domain): implement new Navigator2 and NavigatorJgraphtAdapter2 domains

- calculate shortest path after object initialization
- Navigator<V, E> to specify Vertex and Edge types

bugoverdose committed:
refactor(domain): replace old Navigator and NavigatorJgraphtAdapter with new domains

- fix broken tests

bugoverdose commented on src/main/java/wooteco/subway/service/PathService.java:
```diff
@@ -20,12 +26,15 @@ public PathService(SubwayRepository subwayRepository, StationRepository stationR
+        List<Section> sections = subwayRepository.findAllSections();
 
-        return DtoAssembler.assemble(optimalPath);
+        Path path = Path.of(startStation, endStation, sections);
+        int distance = path.getDistance();
+        List<LineInfo> lines = subwayRepository.findAllLinesByIds(path.getPassingLineIds());
+        Fare fare = new AgeDiscountFare(new LineOverFare(new DistanceOverFare(new BasicFare(), distance), lines), age);
+        return PathResponse.of(path, fare.calculate());
```
요금 계산의 경우, 데코레이터 패턴을 구현해봤습니다. 제대로 구현한게 맞으려나요?

각 추가비용 및 할인 정책마다 필요로 하는 정보가 달라서, 구현체마다 생성자의 인자로 받게 했더니, 예상보다 클라이언트 코드가 더러워졌고, 이름 짓는 것도 어떻게 하는게 적절한지 잘 안떠오르네요.

<img width="603" alt="스크린샷 2022-05-22 오전 2 00 22" src="https://user-images.githubusercontent.com/73531614/169661932-9982d4cd-a236-4d1d-8880-41d61af15383.png">

bugoverdose commented on src/main/java/wooteco/subway/domain/path/Navigator.java:
```diff
@@ -0,0 +1,8 @@
+
+import java.util.List;
+
+public interface Navigator<V, E> {
+
+    List<E> calculateShortestPath(V source, V target);
+}
```
어댑터 패턴을 구현했는데 제대로 구현한건지 좀 의문이 드네요. 

일단 vertex와 edge를 제네릭으로 받는 Navigator 인터페이스를 사용하여, 
두 vertex를 기준으로 가장 가중치의 합이 작은 간선들의 리스트를 반환하도록 했습니다. 

cf) 네비게이터의 경우 Computer 같은 느낌이 들어서 -er/-or로 끝나도 되지 않나 싶었습니다.

<img width="320" alt="스크린샷 2022-05-22 오전 2 00 13" src="https://user-images.githubusercontent.com/73531614/169662119-569973b8-bb74-4bea-bc5c-4e1ddf75e685.png">

bugoverdose replied:
`NavigatorJgraphtAdapter` 객체를 `Path` 외부에서 생성하고, `Path`는 매개변수로 `Navigator` 타입을 받도록 수정했습니다

<img width="377" alt="image" src="https://user-images.githubusercontent.com/73531614/169678536-aad4e67f-239d-4cc9-bc90-0e822070403b.png">

bugoverdose commented on src/main/java/wooteco/subway/domain/path/NavigatorJgraphtAdapter.java:
```diff
@@ -0,0 +1,115 @@
+        return path;
+    }
+
+    private static class SectionEdge extends DefaultWeightedEdge {
+
+        final Section section;
+
+        SectionEdge(Section section) {
+            this.section = section;
+        }
```
Section 자체가 DefaultWeightedEdge를 상속받으면 어댑터 패턴을 쓴게 무의미해진다고 느껴져서 내부 클래스로 Section을 감싸는 DefaultWeightedEdge 구현체를 만들었습니다.

SectionEdge 형식으로 사용하여 최단거리 계산 후에 최종적으로 다시 Section으로 변환하여 외부에 반환되도록 했어요.

bugoverdose reviewed (commented):
Fare는 데코레이터 패턴을 적용했고, jgrapht에 대해서는 어댑터 패턴을 적용해봤습니다.
처음 해본거라 제대로 한건지 잘 모르겠네요.

클래스 이름들을 어떻게 지으면 좋을지 감이 잘 안잡히네요.

시간 되실 때 피드백 바랍니다. :)

bugoverdose committed:
refactor(domain): remove creation of adapter object from Path

- make Path depend only on abstract type Navigator

kang-hyungu commented:
실패하는 테스트 코드를 다시 점검해봅시다

<img width="423" alt="image" src="https://user-images.githubusercontent.com/15669435/169932194-e6340bfd-946c-4007-b4dc-e0c81eb8dc34.png">

kang-hyungu commented on src/main/java/wooteco/subway/domain/fare/AgeDiscountFare.java:
```diff
@@ -0,0 +1,38 @@
+
+public class AgeDiscountFare extends Decorator {
+
+    private static final int BABY_CHILD_AGE_THRESHOLD = 6;
```
지금은 청소년, 어린이 요금만 있지만 새로운 할인 정책이 추가되면 계속 상수를 추가해야되는 구조네요.
enum을 활용해보면 어떨까요?

bugoverdose replied:
외부로 빼기 싫어서 고민하다가 `AgeDiscountFare` 클래스 내부에 `Age`라는 private한 enum을 만들었습니다.

kang-hyungu commented on src/main/java/wooteco/subway/domain/fare/AgeDiscountFare.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    @Override
+    public int calculate() {
```
마찬가지로 새로운 할인 조건이 추가되면 if절이 계속 추가되는 구조네요.
if가 계속 추가되는 구조를 어떻게 바꿀 수 있을까요?

kang-hyungu commented on src/main/java/wooteco/subway/domain/fare/DistanceOverFare.java:
```diff
@@ -0,0 +1,47 @@
+
+public class DistanceOverFare extends Decorator {
+
+    private static final int OVER_FARE_DIGIT = 100;
```
AgeDiscountFare와 마찬가지로 개선할 방법이 없을지 고민해봅시다.

kang-hyungu commented on src/main/java/wooteco/subway/domain/fare/Decorator.java:
```diff
@@ -0,0 +1,14 @@
+package wooteco.subway.domain.fare;
+
+public abstract class Decorator implements Fare {
```
1. 데코레이터라는 클래스명이 지하철과 어떤 관련이 있는건가요?
2. 구현한 코드를 봤을 땐 추상 클래스로 구현할 이유가 잘 안보이네요. 어떤 문제를 해결하기 위해 추가한 클래스인가요?

bugoverdose replied:
1. 좋은 이름이 안떠오르네요. 할인만 있는게 아니라 추가 요금도 있어서 이를 통합하는 용어를 떠올리려 했는데 잘 안되네요.
2. 말로 설명하기 힘든데, 전 이번 미션에서 1단계는 기본 요금에 대해 거리에 따른 추가 요금의 **적용 여부**가 추가된 것이고, 2단계에서는 건너오면서 나이에 따른 할인과 경유한 노선에 따른 추가 요금의 **"적용 여부"가 순차적으로 추가**되었다고 느꼈어요. 

그리고 요금(`Fare`)에 대해 특정한 연산을 추가적으로 적용하는 역할을 `Decorator`가 맡도록 하고 싶었어요. `Decorator`의 각 구현체들이 무조건 연산을 적용할 대상인 `Fare`를 생성자로 받도록 하고 싶었고, 그렇게 만들기 위해 `Fare target`이라는 필드를 지닌 추상클래스로 만들어야 했습니다.

이 구조를 쓰면 요금 계산에 대해 요구사항이 변경되었을 때 `Decorator` 구현체의 적용 순서를 수정하거나, `Decorator` 구현체 자체를 수정하는 등의 방식으로 변화에 유연하게 대처할 수 있다고 느꼈습니다. 

- 1단계 요구사항 : 기본요금 => 거리에 따른 추가 비용 적용
```java
Fare fare = new DistanceOverFare(new BasicFare(), distance);
```

- 2단계 요구사항 추가 : 기본 요금 => 거리에 따른 추가 비용 적용 => 노선에 따른 추가 비용 적용 => 나이에 따른 할인
```java
Fare fare = new AgeDiscountFare(new LineOverFare(new DistanceOverFare(new BasicFare(), distance), lines), age);
```

- 대응할 수 있는 요구사항 변경(적용 순서 변형) : 기본 요금 => 거리에 따른 추가 비용 적용 => 나이에 따른 할인 => 노선에 따른 추가 비용 적용
```java
Fare fare = new LineOverFare(new AgeDiscountFare(new DistanceOverFare(new BasicFare(), distance), age), lines);
```

kang-hyungu commented on src/main/java/wooteco/subway/domain/line/LineInfo.java:
```diff
@@ -4,18 +4,30 @@
 
 public class LineInfo {
```
클린코드
p32
클래스 이름
클래스 이름과 객체 이름은 명사나 명사구가 적합하다. Customer, WikiPage, Account, AddressParser 등이 좋은 예다. Manager, Processor, Data, Info 등과 같은 단어는 피하고, 동사는 사용하지 않는다.

kang-hyungu replied:
클린코드 25p

의미 있게 구분하라
컴파일러나 인터프리터만 통과하려는 생각으로 코드를 구현하는 프로그래머는 스스로 문제를 일으킨다.
예를 들어, 동일한 범위 안에서는 다른 두 개념에 같은 이름을 사용하지 못한다.
그래서 프로그래머가 한쪽 이름을 마음대로 바꾸고픈 유혹에 빠진다.
어떤 프로그래머는 철자를 살짝 바꿨다가 나중에 철자 오류를 고치는 순간 컴파일이 불가능한 상황에 빠진다.(예를 들면 변수명을 klass로 짓는 경우)
컴파일러를 통과할지라도 연속된 숫자를 덧붙이거나 불용어(noise word)를 추가하는 방식은 적절하지 못하다.
이름이 달라야 한다면 의미도 달라져야 한다.
연속적인 숫자를 덧붙인 이름(a1, a2, ... aN)은 의도적인 이름과 정반대다.
이런 이름은 그릇된 정보를 제공하는 이름도 아니며, 아무런 정보를 제공하지 못하는 이름일 뿐이다.
저자 의도가 전혀 드러나지 않는다.
다음 코드를 살펴보자.

```java
public static void copyChars(char a1[], char a2[]) {
    for (int i = 0; i < a.length; i++) {
        a2[i] = a1[i];
    }
```
함수 인수 이름으로 source와 destination을 사용한다면 코드 읽기가 훨씬 더 쉬워진다.

불용어를 추가한 이름 역시 아무런 정보도 제공하지 못한다.
Product라는 클래스가 있다고 가정하자.
다른 클래스를 ProductInfo 혹은 ProductData라 부른다면 개념을 구분하지 않은 채 이름만 달리한 경우다.
Info나 Data는 a, an, the와 마찬가지로 의미가 불분명한 불용어다.
a나 the와 같은 접두어를 사용하지 말라는 소리가 아니다.
의미가 분명히 다르다면 사용해도 무방하다. 예를 들어, 모든 지역 변수는 a를 사용하고 모든 함수 인수는 the를 사용해도 되겠다.
요지는 zork라는 변수가 있다는 이유만으로 theZork라 이름 지어서는 안 된다는 말이다.

kang-hyungu replied:
LineInfo라는 클래스로 따로 분리한 이유가 무엇인가요?
Line 클래스에서 관리해도 되는 데이터들 아닌가요?
오히려 클래스를 분리해서 복잡도를 높이는 건 아닐까요?

bugoverdose replied:
음. Line 클래스는 사실 "노선도"를 의미하고, LineInfo 클래스가 "노선" 그 자체를 지칭한다고 생각해요.
LineInfo라는 이름이 이상한건 계속 인지하고 있어서 확실히 고쳐야 하는데, 두 클래스를 하나로 합쳐야만 하는지는 계속 갈등이 되네요.

물론 Line 클래스 하나만으로 노선에 대한 모든 것을 관리하는 것도 가능하고, 도메인 구조만 생각하면 오히려 그게 더 코드의 가독성도 올라갈 것 같다는 생각도 드는데요.
예를 들어 "노선"에 대한 정보가 필요한 경우, 특정 노선에 등록된 "구간들"에 대한 정보가 필요한 경우, 그 둘이 합쳐진 "노선도"가 통째로 필요한 경우는 구분된다는 느낌이 들어서 지금 구조를 유지했어요.

kang-hyungu commented on src/main/java/wooteco/subway/domain/path/Navigator.java:
```diff
@@ -0,0 +1,8 @@
+
+import java.util.List;
+
+public interface Navigator<V, E> {
```
네비게이터라는 인터페이스가 필요한가요?

bugoverdose replied:
음 기본적으로 Navigator 없이 Path가 NavigatorJgraphtAdapter에 직접 의존하는 것도 불가능하지는 않다고 생각해요.  

기본적으로 NavigatorJgraphtAdapter가 필요한 이유는 Path가 jgrapht 라이브러리에 직접적으로 의존하지 않도록 하기 위함인데요. 만일 jgrapt 이외의 다른 라이브러리를 사용하거나, 이전처럼 다익스트라 알고리즘을 직접 구현한 새로운 Adapter 객체를 만들어서 Path에 넘겨준다고 했을 때, 
1) 해당 객체가 만일 기존 NavigatorJgraphtAdapter 객체와 정확히 동일한 퍼블릭 메서드들을 제공하고 있다면, Path는 NavigatorJgraphtAdapter를 전달받는 부분만 수정하고, 그 이외의 로직은 수정할 필요가 없게 된다고 생각해요.
2) 그런데 만일 NavigatorJgraphtAdapter와 다른 퍼블릭 메서드들을 제공한다면, Path 내부의 로직을 수정해야 할 필요가 생길 것 같아요.

Path가 Navigator라는 추상화에 대해 의존해야 하는 이유는 2)의 경우를 방지하고, 1)을 강제하기 위함이라고 생각해요. 그러니까 모든 Navigator 구현체는 언제나 동일한 퍼블릭 인터페이스를 공유하는데, Path에서 Navigator라는 추상화에 의존하게 되면 다른 adapter 객체를 전달받게 되더라도 Path의 그 어떤 부분도 수정할 필요가 없다고 생각해요.

사실 한편으로는 향후 다른 Navigator 구현체를 만들 일이 없다면, (그러니까 실제로 변경이 일어나지 않는다면,) 변경에 유연하게 대처할 수 있는 구조가 애초에 의미가 없다고 생각해요. 변경에 유연한 구조란 사실 구조가 복잡하다는 의미이기도 해서요.

kang-hyungu commented on src/main/java/wooteco/subway/domain/path/NavigatorJgraphtAdapter.java:
```diff
@@ -0,0 +1,115 @@
+        this.dijkstraPath = dijkstraPath;
+    }
+
+    public static NavigatorJgraphtAdapter of(List<Section> sections) {
```
팩터리 메서드를 따로 만들 이유가 있나요?
생성자에서 충분히 처리 가능한 로직 아닌가요?

kang-hyungu commented on src/main/java/wooteco/subway/domain/path/NavigatorJgraphtAdapter.java:
```diff
@@ -0,0 +1,115 @@
+                .collect(Collectors.toList());
+    }
+
+    private GraphPath<Station, SectionEdge> toValidShortestPath(Station source, Station target) {
```
optional을 사용하면 메서드를 2줄로 줄일 수 있겠네요

```java
    private GraphPath<Station, SectionEdge> toValidShortestPath(Station source, Station target) {
        return Optional.ofNullable(dijkstraPath.getPath(source, target))
                .orElseThrow(() -> new IllegalArgumentException(PATH_NOT_CONNECTED_EXCEPTION));
    }
```

kang-hyungu commented on src/main/java/wooteco/subway/domain/path/Path.java:
```diff
@@ -1,64 +1,54 @@
-
-    public int getDistance() {
-        return distance;
+    public static Path of(Station source, Station target, Navigator<Station, Section> navigator) {
```
왜 팩토리 메서드를 추가했나요?
생성자가 private이지만 Path 클래스 내부에서 사용될 수 있는데 그때는 유효성 검사를 하지 않아서 의도하지 않은 객체가 사용될 수도 있겠네요.

bugoverdose replied:
음 일단 원래는 좀 복잡한 로직이라 당연히 팩토리 메서드가 있어야 한다는 생각이 들었는데, 외부에서 Navigator를 받도록 수정하면서 많이 간결해졌네요.

기본적으로 모든 필드들을 그대로 받는 생성자를 하나 두는 것을 습관화했는데,
`Path`의 경우 `route`의 첫번째 구간에 `source`이 등록되어있어야 한다거나, `route`가 서로 이어지는 구간들이어야 하는 등의 제약조건 때문에 함부로 생성되면 안된다는 생각도 들기는 하네요.

kang-hyungu commented on src/main/java/wooteco/subway/entity/LineEntity.java:
```diff
@@ -8,15 +8,17 @@ public class LineEntity {
+    private final int extraFare;
 
-    public LineEntity(Long id, String name, String color) {
+    public LineEntity(Long id, String name, String color, int extraFare) {
```
엔티티 클래스와 도메인 클래스를 따로 두는게 좋은 구조일까요?
필드를 추가하거나 삭제하는 변경사항이 생길 때 두 클래스를 모두 챙겨야하는데 깜빡하고 놓칠 수도 있겠네요
repository에서 조회한 결과를 도메인 클래스에 바로 매핑해도 되지 않나요?

bugoverdose replied:
엔티티를 두지 않을 수도 있다는 발상을 아예 못하고 있었네요.
너무 당연히 하나의 dao는 하나의 entity에 대응된다고 생각했었는데, 실제로 이번에 extraFare라는 컬럼을 추가하기 위해 dto, domain, entity를 전부 수정해야 했을 때 확실히 번거롭다는 느낌은 받았어요.

그래도 domain 객체들이 dao와 db에 대해 모르게 하는게 유연한 설계를 가능하게 하지 않나?라는 생각이 들어서 어떻게 하는게 맞는지 확신이 서지는 않네요. 이전 체스 미션 때처럼 DB에 저장되는 데이터의 도메인 객체와 동일하지 않은 경우도 분명 존재한다고 생각하고요. 오브젝트에서 유연한 설계는 복잡한 설계를 의미한다는 문구를 읽었는데, 그런 상황이 아닌가 하는 생각도 드네요.

kang-hyungu commented on src/main/java/wooteco/subway/repository/SubwayRepository.java:
```diff
@@ -32,6 +32,12 @@ public List<LineInfo> findAllLines() {
                 .collect(Collectors.toList());
     }
 
```
SubwayRepository 대신에 LineRepository, SectionRepository로 분리해도 충분하지 않을까요?
두 엔티티를 한 곳에서 관리하면 점점 코드가 복잡해질 수 있습니다.

bugoverdose replied:
넵 분리하니까 더 관리하기 좋은 것 같아요.

kang-hyungu commented on src/test/java/wooteco/subway/domain/path/PathTest.java:
```diff
@@ -0,0 +1,89 @@
+@SuppressWarnings("NonAsciiCharacters")
+class PathTest {
+
+    private final Station STATION1 = new Station(1L, "역1");
```
클린코드 25p

의미 있게 구분하라
컴파일러나 인터프리터만 통과하려는 생각으로 코드를 구현하는 프로그래머는 스스로 문제를 일으킨다.
예를 들어, 동일한 범위 안에서 다른 두 개념에 같은 이름을 사용하지 못한다.
...

컴파일러를 통과할지라도 연속된 숫자를 덧붙이거나 불용어(noise word)를 추가하는 방식은 적절하지 못하다.
이름이 달라야 한다면 의미도 달라져야 한다.

연속적인 숫자를 덧붙인 이름(a1, a2 ... aN)은 의도적인 이름과 정반대다.
이런 이름은 그릇된 정보를 제공하는 이름도 아니며, 아무런 정보를 제공하지 못하는 이름일 뿐이다.
저자 의도가 전혀 드러나지 않는다.


클린코드 157p

테스트 코드는 실제 코드 못지 않게 중요하다. 테스트 코드는 이류 시민이 아니다.
테스트 코드는 사고와 설계와 주의가 필요하다.
실제 코드 못지 않게 깨끗하게 짜야한다.

테스트는 유연성, 유지보수성, 재사용성을 제공한다. 테스트 케이스가 있으면 변경이 쉬워지기 때문이다.
따라서 테스트 코드가 지저분하면 코드를 변경하는 능력이 떨어지며 코드 구조를 개선하는 능력도 떨어진다.
테스트 코드가 지저분할수록 실제 코드도 지저분해진다.
결국 테스트 코드를 잃어버리고 실제 코드도 망가진다.

bugoverdose replied:
음. 전 기본적으로 id가 1L인 지하철역을 의미하고자 STATION1이라고 작성했는데, 그러면 오히려 해당 데이터에 대해 필요한 의미를 부여한게 아닌가 싶어요.

<img width="629" alt="image" src="https://user-images.githubusercontent.com/73531614/170046958-165d73fd-c084-4144-8ecf-3f6b67c23c35.png">

<img width="762" alt="스크린샷 2022-05-24 오후 10 21 45" src="https://user-images.githubusercontent.com/73531614/170046264-3cc36acc-43cf-4027-bf3c-ed2a7e348877.png">

예를 들어 `new Station(1L, "강남역")`에게 "강남역"이라는 변수명을 주는 경우 오히려 테스트 내용을 파악하기 어려워지지 않나 싶어요.

<img width="605" alt="스크린샷 2022-05-24 오후 10 22 56" src="https://user-images.githubusercontent.com/73531614/170046083-8d76d5f9-779f-429c-a47a-274b07a98d53.png">

<img width="771" alt="스크린샷 2022-05-24 오후 10 27 02" src="https://user-images.githubusercontent.com/73531614/170046735-62570c8b-553b-44f2-aca8-14b17f703d7d.png">

bugoverdose replied:
근데 section들의 이름은 확실히 중구난방이네요

kang-hyungu commented on src/test/java/wooteco/subway/domain/fare/FareTest.java:
```diff
@@ -0,0 +1,182 @@
+import wooteco.subway.domain.line.LineInfo;
+
+@SuppressWarnings("NonAsciiCharacters")
+class FareTest {
```
하나의 테스트에서 모든 요금을 테스트하지 마시고 각 요금 클래스마다 별도의 테스트를 만들어보세요.
한 클래스에서 모든 조건을 테스트하다보면 코드도 복잡해지고 코드 관리하기 어려워집니다.

kang-hyungu commented on src/test/java/wooteco/subway/test_utils/TestFixtureManager.java:
```diff
@@ -22,8 +22,13 @@ public void saveLine(String name, String color) {
-        saveSection(lineId, upStationId, downStationId,10);
+        saveSection(lineId, upStationId, downStationId, 10);
     }
 
```
test_utils라는 패키지명이 적절한지 우테코 자바 스타일 가이드나 구글 자바 스타일 가이드를 참고해보시기 바랍니다.

bugoverdose replied:
일단 언더스코어(_)를 쓴게 문제군요.

5.2.1 Package names 
Package names use only lowercase letters and digits (no underscores). Consecutive words are simply concatenated together. For example, com.example.deepspace, not com.example.deepSpace or com.example.deep_space.

kang-hyungu reviewed (changes requested):
정 몇 가지 코멘트 남겼어요
수정해보시고 궁금한 점 생기면 코멘트 남기거나 오프라인으로 질문주셔도 좋습니다.

bugoverdose committed:
refactor(test): refactor test setup and fixture classes

- rename test_utils to fixture package
- move http utils out of subway package because it doesn't depend on subway logic
- integrate RepositoryTest, ServiceTest, DaoTest to single DatabaseUsageTest

bugoverdose committed:
refactor(test): separate Fare tests for each decorator

bugoverdose committed:
feat(repository): implement LineRepository and SectionRepository

- separate SubwayRepository to LineRepository and SectionRepository
- add test cases

bugoverdose committed:
refactor(service): replace SubwayRepository with LineRepository and SectionRepository

- remove unnecessary classes

bugoverdose committed:
refactor(path): static factory method at NavigatorJgraphtAdapter to constructor

- fix use cases

bugoverdose committed:
refactor(fare): cleanse AgeDiscountFare logic with inner enum Age

bugoverdose committed:
refactor(domain): rename Line to LineMap

bugoverdose committed:
refactor(domain): rename LineInfo to Line

- because -Info is inappropriate

bugoverdose committed:
refactor(domain): rename Lines to SubwayMap

bugoverdose committed:
refactor(fare): cleanse DistanceOverFare logic with inner enum

bugoverdose committed:
refactor(path): change static factory method to constructor at Path

- only one constructor with validations

bugoverdose committed:
refactor(fare): rename Decorator to FarePolicy

- add abstract or final to all Fare implementations

bugoverdose committed:
refactor(section): rename SectionsCompareResult to SectionUpdates

- minor cleansing

bugoverdose committed:
refactor(test): give more meaning to section variable and constant names

bugoverdose committed:
refactor(fare): cleanse age constants for Age

bugoverdose committed:
refactor(entity): replace StationEntity with responding Station domain

- remove unnecessary StationEntity

bugoverdose committed:
refactor(entity): replace SectionEntity with responding Section domain

- remove unnecessary SectionEntity
- add getters at Section for usage in SectionDao

bugoverdose committed:
refactor(entity): replace LineEntity with responding Line domain

- remove entity package

bugoverdose committed:
refactor(path): make Path depend directly on Adapter concrete class

- remove Navigator interface and rename Adapter to Navigator

bugoverdose committed:
refactor(test): change method signature for saveLine and saveSection at DatabaseFixtureUtils

- generate test fixtures with domains, not specific column values

bugoverdose committed:
refactor(test): change variable and constant station names

- differentiate each station with names not id

bugoverdose committed:
refactor(dto): remove request dto constructor used only for tests

- add private methods at tests

bugoverdose committed:
refactor(test): rename variables and constants

bugoverdose committed:
refactor(line): hide existance of Line inside LineMap

bugoverdose committed:
refactor(repository): make LineRepository return LineMap on findExistingLine

- fix use cases

bugoverdose committed:
refactor(repository): make LineRepository update line data with LineMap

- fix use cases

bugoverdose committed:
refactor: calculate line extra fare with only extra fares, not lines

- fix all use cases

bugoverdose committed:
refactor(repository): make LineRepository save line data with LineMap

- fix use cases

bugoverdose committed:
feat(line): implement LineExtraFare as value object

- move validation from Line to LineExtraFare
- replace int type to LineExtraFare for extraFare field at Line

bugoverdose committed:
refactor: replace all extraFare with LineExtraFare

bugoverdose committed:
refactor(repository): make LineRepository return SubwayMap on findAllLines

bugoverdose committed:
refactor(section): add constructor at Sections

bugoverdose committed:
feat(repository): implement new findAllLines method for returning List<LineMap>

- add test case

bugoverdose committed:
refactor: replace findAllLines with new method

- erase unnecessary SubwayMap class and tests

bugoverdose committed:
refactor(line): remove Line field from LineMap

- fix use cases

bugoverdose committed:
feat(entity): implement LineEntity to replace unnecessary Line domain

bugoverdose committed:
refactor(dao): replace Line with LineEntity

- remove unnecessary Line domain

bugoverdose committed:
refactor(line): rename LineMap to Line

- cleanse constructors

bugoverdose committed:
refactor(Section): remove unused method

- erase unnecessary test cases

bugoverdose committed:
feat(fare): implement new Fare and Policy domains without decorators

- add test cases

bugoverdose committed:
feat(path): implement calculateFare at Path

- add test cases

bugoverdose committed:
refactor(path): replace old fare with new fare package

bugoverdose committed:
feat(dao): batch update for saving or deleting multiple sections

bugoverdose committed:
refactor(test): save test fixtures with domains and entities

- implement batch update

kang-hyungu reviewed (changes requested):
전체 테스트를 돌리니 실패한 테스트가 남아있네요.
미션 제출하기 전에 모든 테스트 코드를 점검하시기 바랍니다.
피드백은 모든 테스트 통과하는 걸 확인하고 남길게요

bugoverdose commented:
어떤게 실패하는지 예외 메시지라도 보여주실 수 있나요?

<img width="505" alt="image" src="https://user-images.githubusercontent.com/73531614/170407087-9ce84606-a380-467f-9942-759393f22d1c.png">

bugoverdose committed:
fix(dto): add default constructor and setters for jackson at acceptance tests

bugoverdose commented on src/test/java/wooteco/subway/acceptance/LineAcceptanceTest.java:
```diff
@@ -199,44 +272,36 @@ class DeleteLineTest {
-            put("name", name);
-            put("color", color);
-        }};
+    private LineResponse extractSingleLineResponseBody(ExtractableResponse<Response> response) {
+        return response.jsonPath().getObject(".", LineResponse.class);
     }
```
프로덕션 코드에서는 response dto들을 직접 생성하니까 괜찮은데,
인수테스트는 json으로부터 LineResponse를 직접 추출하는 로직을 지녀서 디폴트 생성자+setter들을 필요로 하는걸로 추측되네요.

제 컴퓨터에서는 없어도 돌아가는 이유가 gradle 버전이 7.3이라 그런가 싶은데 관련 자료가 잘 안보이네요.

bugoverdose reviewed (commented):
일단 추측되는걸 해봤는데 여전히 실패하면 알려주세요

kang-hyungu commented on src/main/java/wooteco/subway/dao/LineDao.java:
```diff
@@ -43,6 +44,13 @@ public Optional<LineEntity> findById(Long id) {
     }
 
+    public List<LineEntity> findAllByIds(List<Long> ids) {
+        final String sql = "SELECT * FROM line WHERE id in (:ids)";
```
[SELECT * 쓰지 말자](https://hyewoncc.github.io/2022/05/25/sql-no-wildcard.html)

kang-hyungu commented on src/main/java/wooteco/subway/dao/SectionDao.java:
```diff
@@ -68,20 +67,18 @@ public List<SectionEntity> findAllByStationId(Long stationId) {
     }
 
-    public void save(SectionEntity sectionEntity) {
+    public void save(List<Section> sectionEntities) {
```
```suggestion
    public void save(List<Section> sections) {
```

kang-hyungu commented on src/main/java/wooteco/subway/domain/fare/AgeDiscountPolicy.java:
```diff
@@ -0,0 +1,50 @@
+    CHILD(Constants.CHILD_MIN_AGE, Constants.ADOLESCENT_MIN_AGE, 350, 0.50),
+    ADOLESCENT(Constants.ADOLESCENT_MIN_AGE, Constants.ADULT_MIN_AGE, 350, 0.20),
+    ADULT(Constants.ADULT_MIN_AGE, Constants.ELDERLY_MIN_AGE, 0, 0.00),
+    ELDERLY(Constants.ELDERLY_MIN_AGE, 150, 0, 1.00),
```
Constants 라는 inner class로 상수를 관리하려면 0, 150도 추가해주는게 좋지 않을까요?

bugoverdose replied:
좋네요.

```java
public enum AgeDiscountPolicy {

    BABY(Age.MIN, Age.CHILD_MIN, 0, 1.00),
    CHILD(Age.CHILD_MIN, Age.ADOLESCENT_MIN, 350, 0.50),
    ADOLESCENT(Age.ADOLESCENT_MIN, Age.ADULT_MIN, 350, 0.20),
    ADULT(Age.ADULT_MIN, Age.ELDERLY_MIN, 0, 0.00),
    ELDERLY(Age.ELDERLY_MIN, Age.MAX, 0, 1.00),
    ;

...
    private static class Age {

        static final int MIN = 0;
        static final int CHILD_MIN = 6;
        static final int ADOLESCENT_MIN = 13;
        static final int ADULT_MIN = 19;
        static final int ELDERLY_MIN = 65;
        static final int MAX = 150;
    }
}
```

kang-hyungu commented on src/main/java/wooteco/subway/domain/fare/AgeDiscountPolicy.java:
```diff
@@ -0,0 +1,50 @@
+    ELDERLY(Constants.ELDERLY_MIN_AGE, 150, 0, 1.00),
+    ;
+
+    private static final String INVALID_AGE_RANGE_EXCEPTION = "0과 150 사이의 연령만 입력가능합니다.";
```
```suggestion
    private static final String INVALID_AGE_RANGE_EXCEPTION = Constants.MIN_AGE + "과 " + Constants.MAX_AGE + " 사이의 연령만 입력가능합니다.";
```

상수로 관리하면 요구사항이 바꼈을 때 메시지도 놓치지 않고 변경할 수 있겠네요

bugoverdose replied:
<img width="580" alt="image" src="https://user-images.githubusercontent.com/73531614/170489330-cd4a6d32-450c-498f-b088-fed083f13a75.png">

kang-hyungu commented on src/main/java/wooteco/subway/domain/fare/DistanceOverFarePolicy.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    private int calculateOverFare(int overDistance, int limit) {
+        double overFareDigit = Math.ceil((overDistance - 1) / limit) + 1;
```
린트에서 경고가 뜨는데 다른 방법은 없을까요?

<img width="640" alt="image" src="https://user-images.githubusercontent.com/15669435/170449121-aec267fb-4d65-4874-a774-74f7eb7bbc99.png">

<img width="776" alt="image" src="https://user-images.githubusercontent.com/15669435/170449202-7fb83b8a-995c-4c50-92ed-6c01c664da77.png">

bugoverdose replied:
SonarLint 좋네요.

<img width="487" alt="image" src="https://user-images.githubusercontent.com/73531614/170491827-2a17070a-8655-4b72-b604-252a22e42da0.png">

kang-hyungu commented on src/main/java/wooteco/subway/domain/path/Path.java:
```diff
@@ -1,64 +1,61 @@
-    public int hashCode() {
-        return Objects.hash(stations, distance, fare);
-    }
+    public int calculateFare(List<LineExtraFare> extraFares, int age) {
```
fare 변수를 새로 교체할 수도 있지만 요런식으로 써볼 수도 있겠네요

```suggestion
    public int calculateFare(List<LineExtraFare> extraFares, int age) {
        return new Fare()
                .applyDistanceOverFarePolicies(getDistance())
                .applyMaximumLineExtraFare(extraFares)
                .applyAgeDiscountPolicy(age)
                .toInt();
    }
```

bugoverdose replied:
그러네요. 데코레이터 패턴으로 추구했던 것들이 전부 메서드 체이닝으로 되는거였네요.

kang-hyungu reviewed (approved):
정 피드백 반영 잘해주셨어요
코드가 훨씬 읽기 좋고 깔끔해졌네요
추가 코멘트는 다음 미션 진행하면서 반영해보시기 바랍니다.
